### ISSUE TITLE
VSTest Task doesn't stop after test session times out

### ISSUE THREAD

**Issue by compuguru** [Microsoft/vsts-tasks#7318]
## Environment
- Server - VSTS
  - Account Name - PBIX
  - Team Project Name - PowerBIClients
  - Build Definition Name/Build number - Desktop E2E Tests v2 (Custom) / desktop_e2e_v2_custom_20180529.3
   
- Agent - Private
  - Version - 2.133.3
  - OS - Windows Server 2012 R2
 
## Issue Description

We have some UI tests that run on a machine. The agent is setup to run interactively. The build definition that installs the tests first runs the "Visual Studio Test Platform Installer" task to install the framework to run the tests, then runs the "Visual Studio Test" task to run the tests.

The "Visual Studio Test" task has a `.runsettings` file with a `TestSessionTimeout` of 1 hour. The task itself has a timeout of 1.5 hours.  When the `TestSessionTimeout` is hit, tests are stopped running but the task doesn't finish. It waits until its timeout has elapsed and then stops the task.  This results in longer build times and the partial test results are not uploaded to the server.

Any ideas why the Visual Studio Test task does not return then the `TestSessionTimeout` is hit?

### Error logs
```
2018-05-29T18:03:19.1354975Z ##[section]Starting: Run Desktop E2E Tests (ImportFromMinerva disabled)
2018-05-29T18:03:19.1385468Z ==============================================================================
2018-05-29T18:03:19.1385969Z Task         : Visual Studio Test
2018-05-29T18:03:19.1386815Z Description  : Run unit and functional tests (Selenium, Appium, Coded UI test, etc.) using the Visual Studio Test runner. Test frameworks that have a Visual Studio test adapter such as xUnit, NUnit, Chutzpah, etc. can also be run. Tests can be distributed on multiple agents using this task (version 2).
2018-05-29T18:03:19.1387625Z Version      : 2.5.12
2018-05-29T18:03:19.1387993Z Author       : Microsoft Corporation
2018-05-29T18:03:19.1388465Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=835764)
2018-05-29T18:03:19.1389032Z ==============================================================================
2018-05-29T18:03:20.1112389Z Run the tests locally using vstest.console.exe
2018-05-29T18:03:20.1113889Z ========================================================
2018-05-29T18:03:20.1123992Z Test selector : Test assemblies
2018-05-29T18:03:20.1127052Z Test assemblies : D:\_work\3\b/DataExplorerBinaries/Tests/PowerBITests.dll
2018-05-29T18:03:20.1133671Z Test filter criteria : Priority=1&(Testcategory=ReportViewTests|Testcategory=SortTests|Testcategory=ReportPageListTests|Testcategory=InsertImageTest|Testcategory=InsertVisualTests|Testcategory=ImportCustomVisualTests|Testcategory=AnalyticsTests|Testcategory=SaveFileTests2)&(Name!=CutCopyPasteRibbonTest&Name!=UndoRedoKeyDownTest&Name!=CopySwitchViewsAndPasteKeyDownTest&Name!=InsertImage&Name!=CopySwitchViewsAndPasteRibbonTest&Name!=InsertTextbox&Name!=FieldListRenameTable&Name!=FieldListRenameColumn&Name!=FieldListDeleteTable&Name!=MapVisualBingAjaxCallCrossOriginException&Name!=UndoRedoImageVisualTest&Name!=ImportFromMinerva)
2018-05-29T18:03:20.1139785Z Search folder : D:\_work\3\s
2018-05-29T18:03:20.1142562Z Run settings file : D:\_work\3\s\tools\RollingDesktopE2E\desktopE2E.runsettings
2018-05-29T18:03:20.1146392Z Run in parallel : false
2018-05-29T18:03:20.1148632Z Run in isolation : false
2018-05-29T18:03:20.1167488Z Path to custom adapters : null
2018-05-29T18:03:20.1169533Z Other console options : null
2018-05-29T18:03:20.1170973Z Code coverage enabled : false
2018-05-29T18:03:20.1173824Z Rerun failed tests: false
2018-05-29T18:03:20.7629169Z VisualStudio version selected for test execution : toolsInstaller
2018-05-29T18:03:20.9213438Z ========================================================
2018-05-29T18:04:29.2111850Z [command]D:\_work\_tool\VsTest\15.8.0-preview-20180510-03\x64\tools\net451\Common7\IDE\Extensions\TestPlatform\vstest.console.exe @D:\_work\_temp\92622971-636a-11e8-a0d6-e77b5ffe9016.txt
2018-05-29T18:04:31.3268963Z Microsoft (R) Test Execution Command Line Tool Version 15.8.0-preview-20180510-03
2018-05-29T18:04:31.3270134Z Copyright (c) Microsoft Corporation.  All rights reserved.
2018-05-29T18:04:31.3270462Z 
....
2018-05-29T18:15:59.1925113Z Passed   DisableSecurityWarningPromptFromWarningDialog
2018-05-29T18:16:27.7871993Z Passed   TestSecurityWarningPromptSettingFromOptionsDialog
2018-05-29T19:04:30.9972418Z Aborting test run: test run timeout of 3600000 milliseconds exceeded.
2018-05-29T19:04:31.1232064Z Results File: D:\_work\3\s\TestResults\vsoagent_TST-TA-PB-G1-02_2018-05-29_18_09_35.trx
2018-05-29T19:04:31.1258731Z 
2018-05-29T19:04:31.1274754Z Total tests: Unknown. Passed: 5. Failed: 2. Skipped: 4.Test Run Aborted.
2018-05-29T19:04:31.1275271Z 
2018-05-29T19:04:31.1350979Z Test execution time: 1.0000 Hours
2018-05-29T19:27:04.3513901Z ##[error]The operation was canceled.
2018-05-29T19:27:04.3557319Z ##[section]Finishing: Run Desktop E2E Tests (ImportFromMinerva disabled)
```

**Comment by prawalagarwal** [comment]
@compuguru Thanks for reporting. Can you share the logs of this run at devops_tools@microsoft.com?

**Comment by prawalagarwal** [comment]
@compuguru were you able to send the logs?

**Comment by compuguru** [comment]
@prawalagarwal Yup, I sent it on May 31st. Subject of the email is "VSTest Task doesn't stop after test session times out #7318"

**Comment by ShreyasRmsft** [comment]
@compuguru are your tests selenium based tests?

**Comment by ShreyasRmsft** [comment]
@compuguru any updates here?

**Comment by compuguru** [comment]
Sorry for the late response. They're not selenium tests. Let me see if I can come up with a reliable repo. If not, I'll close this issue.

**Comment by ShreyasRmsft** [comment]
@compuguru the general trend we have seen here is that the tests either start certain processes (in case of selenium its browsers) or hold on to some resources that don't get closed or killed and vstest.console.exe waits on these child processes on connections to close before exiting thus causing your timeout.

**Comment by ShreyasRmsft** [comment]
@compuguru closing this for now. Please re-open if you get a reliable re-pro.

**Comment by manu4rhyme** [comment]
@prawalagarwal @ShreyasRmsft 

Hi,

I have been facing the same issue, while trying to on board our product to continuous testing using Selenium. Locally all our Selenium tests are passing, however they are failing on AzDev Ops instead.

I’ve tried following this tutorial
https://docs.microsoft.com/en-us/azure/devops/pipelines/test/continuous-test-selenium?view=vsts

But I’m facing few issues:
1)	Unable to finish/terminate VSTest task (it keeps running indefinitely until I add timeout in control options)
2)	Unable to collect video recordings of the test run https://docs.microsoft.com/en-us/azure/devops/test/collect-screenshots-and-video?view=vsts , so cannot debug failures on the pipeline

Clearly, if the VSTest task were to terminate gracefully, it should ideally publish test results including screenshots/video_recordings, so that we could debug why some test cases are failing on remote agent. 

But the VSTest task is not closing, I've already tried multiple timeout values and hosted agent versions but still the same issue. 

I've also mailed multiple instances of my logs during testing, to the alias mentioned above.

I'm stuck on this since almost a week, until i found this recent forum.
